### ISSUE TITLE
Wrong Handling of RPC Control Messages

### ISSUE THREAD

**Issue by ffa500** [ChainSafe/js-libp2p-gossipsub#117]
All released versions of libp2p-gossisub use logical OR in the guarding condition of the function [`_processRpcControlMessage`](https://github.com/ChainSafe/js-libp2p-gossipsub/blob/a590997c6a5f689b1b629fa4348c3b4987c79658/ts/index.ts#L185):

```
_processRpcControlMessage (peer: Peer, controlMsg: ControlMessage): void {
    if (!controlMsg) {
      return
    }

    const iwant = this._handleIHave(peer, controlMsg.ihave)
    const ihave = this._handleIWant(peer, controlMsg.iwant)
    const prune = this._handleGraft(peer, controlMsg.graft)
    this._handlePrune(peer, controlMsg.prune)

    if (!iwant || !ihave || !prune) {
      return
    }

    const outRpc = createGossipRpc(ihave, { iwant: [iwant], prune })
    this._sendRpc(peer, outRpc)
}
```

I assume logical AND should be used instead, as an outbound control message may include any non-empty combination of IWANT, IHAVE, and PRUNE:

```
_processRpcControlMessage (peer: Peer, controlMsg: ControlMessage): void {
    if (!controlMsg) {
      return
    }

    const iwant = this._handleIHave(peer, controlMsg.ihave)
    const ihave = this._handleIWant(peer, controlMsg.iwant)
    const prune = this._handleGraft(peer, controlMsg.graft)
    this._handlePrune(peer, controlMsg.prune)

    if (!iwant && !ihave && !prune) {
      return
    }

    const outRpc = createGossipRpc(ihave, { iwant: iwant && [iwant], prune })
    this._sendRpc(peer, outRpc)
}
```

Note that composing the iwant array in `outRpc` message requires an additional check now.

The [`go-libp2p-gossipsub`](https://github.com/libp2p/go-libp2p-pubsub/blob/99507107b678e875730eef9dec88f8fc9e6c6b5d/gossipsub.go#L513) reference implementation of the corresponding function also uses this logic:

```
func (gs *GossipSubRouter) HandleRPC(rpc *RPC) {
    ctl := rpc.GetControl()
    if ctl == nil {
        return
    }

    iwant := gs.handleIHave(rpc.from, ctl)
    ihave := gs.handleIWant(rpc.from, ctl)
    prune := gs.handleGraft(rpc.from, ctl)
    gs.handlePrune(rpc.from, ctl)

    if len(iwant) == 0 && len(ihave) == 0 && len(prune) == 0 {
        return
    }

    out := rpcWithControl(ihave, nil, iwant, nil, prune)
    gs.sendRPC(rpc.from, out)
}
```

I came across this issue in a use case where IHAVE messages received by a subscribing peer were not answered by IWANT messages which, in consequence, led to lost publication messages.

Thanks a lot for fixing this issue in a patch release!

Hubertus

BTW, the pull request for [`[WIP] Gossipsubv1.1`](https://github.com/ChainSafe/js-libp2p-gossipsub/pull/82) is also affected.

**Comment by vasco-santos** [comment]
Hey @ffa500 
Thanks for the report. We had already fixed this for gossipsub1.1, and I just created PRs to fix this in the previous 2 gossipsub releases.

**Comment by ffa500** [comment]
Great! Thanks a lot!

**Comment by vasco-santos** [comment]
Closing this. A fix was released in `0.4.7`, `0.5.1` and `0.6.0`
